### PR TITLE
fix(entity-resolution): Support keycloak entity resolution for jwt entities

### PR DESCRIPTION
**Pull request by elizabethhealy**
Allow for entity resolution for jwts, get the client/user instead of pulling the subject external value field from the jwt directly

Question to think about: for a clientID, should we return both the client and its service account user (if exists)?

### CODE REVIEW

File: service/internal/idpplugin/keycloak_plugin.go [hunk 1]
```diff
@@ -58,14 +58,47 @@ func EntityResolution(ctx context.Context,
...
+		case *authorization.Entity_Jwt:
+			entityJwt := ident.GetJwt()
+			// decode the jwt
+			_, mapClaims, err := connector.client.DecodeAccessToken(ctx, entityJwt, kcConfig.Realm)
+			if err != nil {
+				return &authorization.IdpPluginResponse{},
+					status.Error(codes.Internal, db.ErrTextCreationFailed)
+			}
+			// if it has client ID use that
+			clientID, ok := (*mapClaims)["client_id"].(string)
```
Comment (pflynn-virtru) [inline service/internal/idpplugin/keycloak_plugin.go:83]: @mkleene you found an oddity with `client-id`, I believe.  Is this correct logic for Keycloak?

Comment (mkleene) [reply service/internal/idpplugin/keycloak_plugin.go:83]: according to this: https://datatracker.ietf.org/doc/html/rfc9068, it's required, but there are lots of cases where it's not included

Comment (mkleene) [reply service/internal/idpplugin/keycloak_plugin.go:83]: and that's a proposed standard so it makes sense

**Review by pflynn-virtru — DISMISSED** [review]

File: service/internal/idpplugin/keycloak_plugin.go [hunk 2]
```diff
@@ -94,95 +127,90 @@ func EntityResolution(ctx context.Context,
...
-					if exErr != nil {
+		} else if getUserParams != (gocloak.GetUsersParams{}) {
+			// if getUserParams not empty
+			users, err := connector.client.GetUsers(ctx, connector.token.AccessToken, kcConfig.Realm, getUserParams)
+			if err != nil {
+				slog.Error(err.Error())
+				return &authorization.IdpPluginResponse{},
+					status.Error(codes.Internal, db.ErrTextGetRetrievalFailed)
+			}
+			if len(users) == 1 {
```
Comment (strantalis) [inline service/internal/idpplugin/keycloak_plugin.go:138]: Is there ever a chance we will somehow get more than 1 user?

File: service/internal/idpplugin/keycloak_plugin.go [hunk 3]
```diff
@@ -94,95 +127,90 @@ func EntityResolution(ctx context.Context,
...
-						entityNotFoundErr = authorization.EntityNotFoundError{Code: int32(codes.NotFound), Message: db.ErrTextGetRetrievalFailed, Entity: ident.GetEmailAddress()}
-					case *authorization.Entity_UserName:
-						entityNotFoundErr = authorization.EntityNotFoundError{Code: int32(codes.NotFound), Message: db.ErrTextGetRetrievalFailed, Entity: ident.GetUserName()}
-					// case "":
-					// 	return &authorization.IdpPluginResponse{},
-					// 		status.Error(codes.InvalidArgument, db.ErrTextNotFound)
-					default:
-						slog.Error("Unsupported/unknown type for", "entity", ident.String())
-						entityNotFoundErr = authorization.EntityNotFoundError{Code: int32(codes.NotFound), Message: db.ErrTextGetRetrievalFailed, Entity: ident.String()}
+					if len(groups) == 1 {
```
Comment (strantalis) [inline service/internal/idpplugin/keycloak_plugin.go:159]: Same here is it possible to get more than one group?

Comment (pflynn-virtru) [reply service/internal/idpplugin/keycloak_plugin.go:159]: it is possible, let's modify the `slog.Error` to "Unexpected group count found for "entity", ident.String(), "count",  len(groups)"

**Pull request closed without merging**